core/shared/src/main/scala/zio/ZSchedule.scala

```diff
@@ -496,8 +496,8 @@ trait ZSchedule[-R, -A, +B] extends Serializable { self =>
    * for every decision of this schedule. This can be used to create schedules
    * that log failures, decisions, or computed values.
    */
-  final def onDecision[A1 <: A](f: (A1, self.State) => UIO[Unit]): ZSchedule[R, A1, B] =
-    updated(update => (a, s) => update(a, s).tap(step => f(a, step)))
+  final def onDecision[A1 <: A, R1 <: R](f: (A1, Option[self.State]) => URIO[R1, Any]): ZSchedule[R1, A1, B] =
+    updated(update => (a, s) => update(a, s).tap(step => f(a, Some(step))))
```

Comment:
Why change to `Option`? It looks like you are always wrapping into `Some`?

The expansion to allow use of `R` in the updater is good.

Reply:
The idea is to `fold` the result of `update` so that a `fail(())` results in `None` and a success results in `Some`

Reply:
@YuvalItzchakov ^^